### DISCUSSION TITLE
Question: motivation for by-index list initialization

### DISCUSSION THREAD

**Discussion by fandrei** [dotnet/csharplang#8898] [Q&A]
For example, in the docs:
```C#
var countdown = new TimerRemaining()
{
    buffer =
    {
        [^1] = 0,
        [^2] = 1,
        [^3] = 2,
        [^4] = 3,
        [^5] = 4,
        [^6] = 5,
        [^7] = 6,
        [^8] = 7,
        [^9] = 8,
        [^10] = 9
    }
};
```
(Let's set aside the fact that this code is awful and should never be used)
However, it's just unrealistic. What are real world examples where this feature is necessary, and why not achieve same results using other language features?

**Comment by HaloFour** [comment]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-11-15.md#evaluation-of-implicit-indexers-in-object-initializers

**Reply by fandrei** [reply]
@HaloFour thanks. But it's a discussion of implementation details rather than explanation of motivation for this feature.

**Reply by jnm2** [reply]
From that link:
> This is more of a bugfix clarification to the range feature in C# 8. 

This wasn't a standalone feature, it was making sure what was implemented was actually what was designed in C# 8. (Or clarifying what was actually meant in the design.)

**Reply by fandrei** [reply]
> Or clarifying what was actually meant in the design

@jnm2 this still doesn't answer "why" :shrug:

**Reply by CyrusNajmabadi** [reply]
I don't see why this wouldn't be supported.  If i want to overwrite a particular element of the collection, then why would i not be able to use indexers to do this?

**Reply by fandrei** [reply]
> I don't see why this wouldn't be supported. If i want to overwrite a particular element of the collection, then why would i not be able to use indexers to do this?

Because it took extra work, and I can't imagine a single real world example where this feature would be necessary.
Can you? :shrug:

**Reply by HaloFour** [reply]
The language team decided it was worth clearing up the bugs in the spec.  They're really the only ones who need to deem whether or not that was necessary to do so.

**Reply by fandrei** [reply]
> The language team decided it was worth clearing up the bugs in the spec.

Sure. But where can I read about the reasons and motivation?

**Reply by jnm2** [reply]
> Because it took extra work, and I can't imagine a single real world example where this feature would be necessary.
Can you? 🤷

It could actually be extra work to make a special case and block these virtual indexers from existing in object initializers. It's not at all clear that what has been landed on is more costly.

**Reply by jnm2** [reply]
Read <https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md> and its linked meeting notes for the motivation for adding virtual indexers to these types.

**Reply by CyrusNajmabadi** [reply]
> Because it took extra work,

It did not.  This falls out from normal indexers in object initializers.  We'd have to go out of our way to *block* this.

> and I can't imagine a single real world example where this feature would be necessary.
> Can you? 🤷

Yes.  As we've discussed in the past, having types that have some amount of precomputed values in their collections is something *normal* to me.  So being able to overwrite any of them makes sense as well.

**Reply by fandrei** [reply]
> It did not. This falls out from normal indexers in object initializers. We'd have to go out of our way to _block_ this.

Nothing had to be blocked, from-the-end indexes simply didn't work in this context until C# 13. :shrug:

> Yes. As we've discussed in the past, having types that have some amount of precomputed values in their collections is something _normal_ to me. So being able to overwrite any of them makes sense as well.

I'd love to see a real-world example of using by-index access being necessary to solve this task. And I mean any type of by-index list initializers, not just from-the-end ones.

**Reply by CyrusNajmabadi** [reply]
> I'd love to see a real-world example of using by-index access being necessary to solve this task.

Nothing is ever necessary.  You can always solve it in different ways.  That's not relevant to me though.  What's relevant is that our language features work in a uniform fashion here.  We have indexed initialization.  You may not find it valuable.  But we think it is.  And, as such, having it work uniformly in these contexts is a good place to be.

**Reply by fandrei** [reply]
> Nothing is ever necessary. You can always solve it in different ways. That's not relevant to me though.

@CyrusNajmabadi  as much as I remember, you said that features that don't provide additional value for the users, or provide value only for a tiny minority of the users, should not be implemented.
What value does this type of initialization provide comparing to existing features that solve the same problem? Can I see a _real world example_ of the problem that it solves?

**Reply by CyrusNajmabadi** [reply]
> What value does this type of initialization provide

Being able to initialize collections at particular indices was felt to be valuable.  

> Can I see a real world example of the problem that it solves?

It's not my job to go hunt down examples for you.   Nor am i interested in doing this for you (esp. given the behaviors experiences so far).  You're welcome to go spend that time if you want.  I'm telling you though that i certainly have use indexer initialization before and found it valuable.

**Reply by fandrei** [reply]
> given the behaviors experiences so far

I only asked for an example. What's wrong with such behavior? :shrug:

**Reply by CyrusNajmabadi** [reply]
I'm referring to your behavior across several issues, which has been highly combative, and has shown an unwillingness to accept differing perspectives.  In other words, it's been unpleasant.  Given that, it hasn't endeared (at least) me, to take large efforts to help you out in areas like this.  

In other words, you attract more bees with honey than vinegar :)

**Reply by fandrei** [reply]
> which has been highly combative

Everything would be much easier if you just gave a real example of the code (that I asked for in the first place), and then we would simply proceed to comparison of different approaches.
And you would have saved a lot of effort if you just did that.

**Reply by CyrusNajmabadi** [reply]
> Everything would be much easier if you just 

This is combative.  

Please stop.  Again, it's in no way my responsibility (or even desire) to go do this work for you.  

If you want people to help you out in specific ways, you'll have ot accept that that involves meeting them at a reasonable discourse level.  We outline this at: https://dotnetfoundation.org/about/policies/code-of-conduct

If you're not willing to engage at this level, then people (myself included) will not be willing to go out of their way to do this work that you are asking.  

*Again*: you attract more bees with honey than vinegar :)

Not sure how to make that any clearer.

--

Note: Joseph already said something similar to you:

> I'm not really attracted by the setup of "here's an idea I'm attached to, and it's going to be strenuous to change my mind." That's work that you can do on your own, without me slowing down your process :)

**Reply by fandrei** [reply]
> This is combative.

You don't have to agree with me, and I don't have to agree with you. Right?

> If you're not willing to engage at this level

I'd prefer the level where we discuss examples of the code rather than personalities, if possible.

**Reply by CyrusNajmabadi** [reply]
> You don't have to agree with me, and I don't have to agree with you. Right?

If you wish to discuss things with people, then behavior matters.

> I'd prefer the level where we discuss examples of the code rather than personalities, if possible.

You're welcome to prefer that.  I prefer discussing things with people who can do so in a pleasant fashion.  C'est la vie.

At this point, multiple people have expressed to you that they're not interested in engaging with you in this fashion.  Ultimately, it's up to you to determine how that leads you.  If you want to ignore that, that's up to you.

**Reply by fandrei** [reply]
> I prefer discussing things with people who can do so in a pleasant fashion.

You spent so much time and effort discussing my behavior that you don't find pleasant enough, rather than just answer the question.
As a side note - I don't like your behavior either, but such matters are just offtopic here.

**Reply by Digitalio1** [reply]
> Being able to initialize collections at particular indices was felt to be valuable.

I'd like to see an example of the code where this feature was necessary, too.

**Reply by CyrusNajmabadi** [reply]
@Digitalio1 The feature is being able to use indexers inside an object initializer.  Something that is nice when you want to do things like `new Dictionary<string, int> { ["foo"] = 1 }`.  The way the feature is spec'ed is general though, and simply involves normal lookup to indexers with normal arguments.  That means that functionality like how indexers work *in a general sense* just fall out here.

So, for example, when we did the work for Index/Range, and the synthesis of implicit index-taking indexers (which you can read about here: https://github.com/dotnet/csharplang/blob/main/proposals/csharp-8.0/ranges.md) then this just fell out in terms of how the language works.

So there was no statement that this specific functionality was necessary.  Simply that we want indexers to work within object initializers.  And that means picking up all the behavior that indexers support anyways.  We'd have to do *more* work to block this.  

This was covered above.  As jnm mentioned:

> It could actually be extra work to make a special case and block these virtual indexers from existing in object initializers. It's not at all clear that what has been landed on is more costly.

Basically, think of the lang spec as a set of rules that combine together to form the totality of the language.  Sometimes those parts combine in interesting ways, and ways that may not be of htat much use to you.  We do not go out of our way to block these things unless we think there is a very strong reason to do so.

In this case, we have indexers available in object initializers for good reason.  And we have synthesized index-taking indexers (again, for good reason).  These happen to combine here in the spec/impl in a way that may not be of much use to you.  But that doesn't mean we're going to block it.  That would take *extra* effort in the spec *and* teh impl to make this *stop* working.  But absent incredible evidence that we should block this (for example, that people are attempting to use this in a widepsread manner, and it's causing all sorts of problems), we'll simply opt for the simpler language spec that defines the two separate features, and is fine if they happen to allow this sort of thing.

**Reply by Digitalio1** [reply]
> That would take _extra_ effort in the spec _and_ teh impl to make this _stop_ working.

That simply is not true. The feature that allows to init values by index was added in C# 3 (as much as I remember), and init by reverse indexes was added in v13. It took extra effort to add these feature, not to block them.
And as much as I remember, you said that language shouldn't be changed unless it's really necessary.

**Reply by HaloFour** [reply]
> The feature that allows to init values by index was added in C# 3 (as much as I remember)

Incorrect, it was added in C# 6.0.  Collection initializers prior to C# 6.0 supported no mechanism to call the indexer, at best you could call an `Add` method, but that could not be mixed with property initializers.

**Reply by Digitalio1** [reply]
> Incorrect, it was added in C# 8.0

Ok, it was added in the different version of the language. My argument still stands. It was _added_, and it took extra effort to add it.

**Reply by CyrusNajmabadi** [reply]
> It took extra effort to add these feature, not to block them.

this is not the case.  The feature was supported *as far as the language was concerned*.  And the impl wants to be simple by not adding rules it doesn't need.  Any work done here was in the efforts of keeping things simple and in line with the langauge spec.  Otherwise:

1. the impl is out of spec (which is costly in its own right for doc'ing and maintaining)
2. the impl is more complex (since it has mistakes in it from not using the standard codepaths already in place for these features).

**Reply by HaloFour** [reply]
And the combination of the spec changes for the two features led to the clarification required here.  Did it require effort to implement?  Sure.  And the language team deemed it worthy of the effort to change the compiler to behave as expected given the existing language specification.  The language team is the arbiter as to whether or not changes are necessary.

**Reply by CyrusNajmabadi** [reply]
> And as much as I remember, you said that language shouldn't be changed unless it's really necessary.

Right.  That's the argument we're making here.  We have rules for indexable collections.  We have rules for synthesized indexers.  We believe these should just work as spec'ed, without special behavior.  The implementation wants to follow that for simplicity of spec and impl.  

We don't change things when we don't think it's necessary.  In the case where it brings everything into alignment and prevents undesirable impl deviations we may change things (or not).

**Reply by Digitalio1** [reply]
> The feature was supported _as far as the language was concerned_.

I mean the original index-init feature, not the later reverse index addition.

**Reply by CyrusNajmabadi** [reply]
> The feature that allows to init values by index was added in C# 3 (as much as I remember), 

We *want* both of these features.  We *want* initialization by indexing (added in C# 6.0 btw).   It allows for `new Dict<string, int> { ["a"] = 1 }.` 

We also want synthesized indexers.  That's what allows you to use things like `foo[^1]` on a type that does not explicitly have an indexer that takes a range (but  has a `this[int x] { get; }` indexer and an `int Length { get; }` property.  This is also a good feature that we want. 

As ssuch the language added these and the impl was updated to support this.  A *consequence* of this is that because there are not synthesized indexers which support System.Index, and because there are initializers, it's the case that you can use System.Index in an initializer.  *That's it*. 

You might not think this is useful or valuable.  That's fine.  But this falls out because of two features we did that we think are super valuable and we want in the language.  

> nd it took extra effort to add it.

No.  I don't think you underrstand how the language works or how the impl works.  The language for initializers says that if you use the `[...] = ...` form that we  lookup the indexers, apply normal overload resolution to them with the supplied argument and assigned value.  The language for indexers then covers how this works with things like System.Index.  So this *falls out* from how the two parts of the spec work.

For this to *not* work we'd have to add language in the spec to *block it*.  That's extra work.  Work we don't think is valuable.  We block when we think something is bad or because we want much more time to think on it.  Not because it doesn't seem that useful.

**Reply by CyrusNajmabadi** [reply]
> I mean the original index-init feature

Yes.  We think the original index-init feature is a *good thing*.  How many more times do you need to hear this?  We think it's a good thing that you can type `new Dictionary<string, int> { ["some key"] = some_value }`.

This also applies to types like ConcurrentDictionary which you cannot use with `{ x, y }`  but which does work with `{ [x] = y }`.

**Reply by CyrusNajmabadi** [reply]
The spec for initializers says:

```
object_initializer
    : '{' member_initializer_list? '}'
    | '{' member_initializer_list ',' '}'
    ;

member_initializer_list
    : member_initializer (',' member_initializer)*
    ;

member_initializer
    : initializer_target '=' initializer_value
    ;

initializer_target
    : identifier
    | '[' argument_list ']'
    ;

initializer_value
    : expression
    | object_or_collection_initializer
    ;
```

There you see the grammar structure that allows the member_initializer target to be `[...]`.  The rules then say:

> An object initializer consists of a sequence of member initializers, enclosed by { and } tokens and separated by commas. Each member_initializer shall designate a target for the initialization. An identifier shall name an accessible field or property of the object being initialized, *whereas an argument_list enclosed in square brackets shall specify arguments for an accessible indexer on the object being initialized.*

Relevant part bolded.  The spec says that you have to have an accessible indexer, and that is what the target of initialization is.  Because of this, indexing with a System.Index argument *falls out*.  Because those indexers are synthesized due to other language in the spec that determines what indexers a type has.

**Reply by Digitalio1** [reply]
> We think the original index-init feature is a _good thing_. How many more times do you need to hear this? We think it's a good thing that you can type `new Dictionary<string, int> { ["some key"] = some_value }`.

What for? It's only new syntax that does the same thing as old syntax. It doesn't add any value. It doesn't let you do something that was difficult without it. It doesn't even save time. The only thing that it adds is the new type of bugs.

> This also applies to types like ConcurrentDictionary

This makes somewhat more sense, but:
1. ConcurrentDictionary is used rarely
2. It's very easy to achieve the same result using features that already existed. No need to create the new language feature just for this rare use case.

**Reply by CyrusNajmabadi** [reply]
> What for? It's only new syntax that does the same thing as old syntax. 

It doesn't do the same thing as old syntax. You can see that in the spec.

**Reply by CyrusNajmabadi** [reply]
>  this rare use case.

Initializing indexable types using their indexers is not a rare case. 

Regardless, we simply do not agree with you. You do not find this feature compelling.  That's your right. We *do*.  So we added it.

The requests for it were something we found reasonable.  The use cases were compelling enough. And the costs were acceptable. 

That's all that matters here. Universal acceptance of a feature is a non goal for us. You not liking it or thinking it is worthwhile literally doesn't impact the decision making. Especially the decision making from literally *2015*. 

> No need to create the new language feature just for

Let me be as blunt as possible, as this discussion does not seem to be adding much value.  This is not your call to make.  If you would not take this feature for your own language *that's your call*. But this is our language, and we take in the data, requests, use cases, etc and we make determinations in what we think is worth investing in. 100% of the features we do have people who do not think it is worthwhile. That simply is not relevant to us. We do the features because *we* believe their value justifies their cost. You're welcome to disagree. But we're not relitigating decisions we made 10 years ago that we still think we're the right call.

**Reply by Digitalio1** [reply]
> The use cases were compelling enough.

I've yet to see a single one that actually offers _any_ improvement. I'm not even talking about necessity.
You've spent an hour beating around the bush, and it looks like "we found it useful, because reasons" is the only answer that you have.

**Reply by CyrusNajmabadi** [reply]
> I've yet to see a single one that actually offers any improvement

So be it.  It's not our job to provide you with cases you find compelling.  As mentioned, with every release there are people like yourself that feel this way about every feature we do.  It's unclear though then why you think this would be a concern for us.  We have no goal around providing suitable examples that convince all users that a feature is worthwhile.  It's a *non goal*.  You keep expecting us to do something which we see no need or value in doing.

> " "we found it useful" ...  is the only answer that you have.

It's the only answer that we need.  That's what language design is.  We make determinations on what we think will be useful for the ecosystem and we design and implement that.

**Comment by CyrusNajmabadi** [comment]
You might prefer coming to our discord to just chat directly about your questions.  You can directly interact live/async at: https://discord.com/invite/tGJvv88

**Reply by fandrei** [reply]
> You might prefer coming to our discord to just chat directly about your questions. You can directly interact live/async at: https://discord.com/invite/tGJvv88

In this case, this info would be lost for the people who will have same question in the future.

**Reply by CyrusNajmabadi** [reply]
No it wouldn't.  Discord is searchable and indexed and people find and comment on threads from well in teh past plenty of times.  :)

It's much more conducive to this form of rapid-fire back and forth  that you're doing here :)

**Comment by CyrusNajmabadi** [comment]
At this point i'm closing and locking this discussion.  There appears to be no interest in engaging with this topic.  The discussion is going in circles, with none of the information provided being accepted.
